File: backend/src/OptimisticColoringAllocator.cpp
#include "OptimisticColoringAllocator.h"
#include "EventEngine.h"
#include <algorithm>
#include <set>

OptimisticColoringAllocator::OptimisticColoringAllocator(InterferenceGraph& g, size_t k)
    : graph(g), K(k) {}

void OptimisticColoringAllocator::allocate() {
    // Simplified Briggs optimistic coloring
    // 1. Simplify
    // 2. If no nodes < K, push node with highest degree anyway (don't spill immediately)
    // 3. Select colors, potentially spilling only if neighborhood is fully colored

    std::vector<std::string> stack;
    std::set<std::string> remaining;
    
    // Create local adjacency list
    std::map<std::string, std::set<std::string>> adj;
    const auto& vars = graph.getVariables();
    for (const auto& v : vars) {
        remaining.insert(v);
        adj[v] = std::set<std::string>();
    }
    
    // Reconstruct edges
    const auto& edges = graph.getEdges();
    for (const auto& e : edges) {
        adj[e.source].insert(e.target);
        adj[e.target].insert(e.source);
    }
    
    while (!remaining.empty()) {
        bool found = false;
        for (const auto& v : remaining) {
            if (adj[v].size() < K) {
                stack.push_back(v);
                for (const auto& neighbor : adj[v]) {
                    adj[neighbor].erase(v);
                }
                remaining.erase(v);
                found = true;
                break;
            }
        }
        
        if (!found) {
            // Optimistic push
            std::string maxDegreeNode = *remaining.begin();
            size_t maxDeg = adj[maxDegreeNode].size();
            for (const auto& v : remaining) {
                if (adj[v].size() > maxDeg) {
                    maxDeg = adj[v].size();
                    maxDegreeNode = v;
                }
            }
            stack.push_back(maxDegreeNode);
            for (const auto& neighbor : adj[maxDegreeNode]) {
                adj[neighbor].erase(maxDegreeNode);
            }
            remaining.erase(maxDegreeNode);
        }
    }
    
    // Rebuild full adjacency
    for (const auto& v : vars) {
        adj[v] = std::set<std::string>();
    }
    for (const auto& e : edges) {
        adj[e.source].insert(e.target);
        adj[e.target].insert(e.source);
    }
    
    // Assign colors
    for (auto it = stack.rbegin(); it != stack.rend(); ++it) {
        std::string v = *it;
        std::set<int> usedColors;
        for (const auto& neighbor : adj[v]) {
            if (assignment.find(neighbor) != assignment.end()) {
                usedColors.insert(assignment[neighbor]);
            }
        }
        
        int color = -1;
        for (size_t c = 0; c < K; ++c) {
            if (usedColors.find(c) == usedColors.end()) {
                color = c;
                break;
            }
        }
        
        if (color == -1) {
            spillCandidates.push_back(v);
        } else {
            assignment[v] = color;
            EventStore::getInstance().emit("ColorAssigned", "allocation", "Optimistic assigned color", v, "{\"color\":" + std::to_string(color) + "}");
        }
    }
}

std::string OptimisticColoringAllocator::toJSON() const {
    std::string json = "{\"K\":" + std::to_string(K) + ",\"spillRequired\":" + (spillCandidates.empty() ? "false" : "true") + ",\"assignment\":{";
    bool first = true;
    for (const auto& kv : assignment) {
        if (!first) json += ",";
        json += "\"" + kv.first + "\":{\"register\":" + std::to_string(kv.second) + ",\"name\":\"R" + std::to_string(kv.second + 1) + "\"}";
        first = false;
    }
    json += "},\"spillCandidates\":[";
    for (size_t i = 0; i < spillCandidates.size(); ++i) {
        json += "\"" + spillCandidates[i] + "\"";
        if (i < spillCandidates.size() - 1) json += ",";
    }
    json += "]}";
    return json;
}

File: frontend/src/lib/OptimisticColoringAllocator.ts
// ============================================================
// OptimisticColoringAllocator.ts — Optimistic Graph Coloring
// ============================================================
import {
  InterferenceGraphData,
  AllocationStep,
  AlgorithmComparisonResult,
} from "@/types"

/**
 * Optimistic Coloring (Briggs variant):
 * Like Chaitin but marks spill candidates WITHOUT removing them immediately.
 * After simplification, it tries to color optimistic candidates too.
 * This often results in fewer actual spills than conservative Chaitin.
 */
export class OptimisticColoringAllocator {
  private graph: InterferenceGraphData

  constructor(graph: InterferenceGraphData) {
    this.graph = graph
  }

  allocate(): AlgorithmComparisonResult {
    const K = this.graph.K
    const regNames = ["R1", "R2", "R3", "R4", "R5", "R6", "R7", "R8"]
    const steps: AllocationStep[] = []
    const spillCandidates: string[] = []
    const assignment: Record<string, { register: number; name: string }> = {}

    // Build adjacency list
    const adjList: Record<string, string[]> = {}
    for (const v of this.graph.variables) {
      adjList[v] = this.graph.edges
        .filter(e => e.source === v || e.target === v)
        .map(e => e.source === v ? e.target : e.source)
    }

    // Working copy
    const tempAdj: Record<string, Set<string>> = {}
    for (const v of this.graph.variables) {
      tempAdj[v] = new Set(adjList[v] || [])
    }

    const stack: string[] = []
    const optimisticCandidates = new Set<string>()
    let remaining = new Set(this.graph.variables)

    const getDeg = (v: string) => tempAdj[v]?.size || 0

    // Simplification — optimistic: always push, mark high-degree as "optimistic"
    while (remaining.size > 0) {
      let found = false

      // Try to find node with degree < K first
      for (const v of remaining) {
        if (getDeg(v) < K) {
          stack.push(v)
          steps.push({
            type: "SIMPLIFY_PUSH",
            variable: v,
            registerId: -1,
            message: `PUSH ${v} (degree ${getDeg(v)} < K=${K})`,
            currentAssignment: {},
            currentStack: [...stack],
            spillCandidates: Array.from(optimisticCandidates),
            remainingGraph: {},
          })
          for (const n of tempAdj[v] || []) tempAdj[n]?.delete(v)
          delete tempAdj[v]
          remaining.delete(v)
          found = true
          break
        }
      }

      if (!found) {
        // Optimistic: push anyway (don't actually spill yet, just mark)
        let maxDeg = -1
        let candidate = ""
        for (const v of remaining) {
          if (getDeg(v) > maxDeg) { maxDeg = getDeg(v); candidate = v }
        }
        if (candidate) {
          optimisticCandidates.add(candidate)
          stack.push(candidate)
          steps.push({
            type: "SIMPLIFY_SPILL",
            variable: candidate,
            registerId: -1,
            message: `OPTIMISTIC PUSH ${candidate} (degree ${maxDeg} >= K=${K}) — will try to color anyway`,
            currentAssignment: {},
            currentStack: [...stack],
            spillCandidates: Array.from(optimisticCandidates),
            remainingGraph: {},
          })
          for (const n of tempAdj[candidate] || []) tempAdj[n]?.delete(candidate)
          delete tempAdj[candidate]
          remaining.delete(candidate)
        } else break
      }
    }

    // Color assignment — pop stack
    const colorMap: Record<string, number> = {}
    const revStack = [...stack].reverse()

    for (const v of revStack) {
      const usedColors = new Set<number>()
      for (const n of adjList[v] || []) {
        if (colorMap[n] !== undefined && colorMap[n] >= 0) usedColors.add(colorMap[n])
      }

      let color = -1
      for (let c = 0; c < K; c++) {
        if (!usedColors.has(c)) { color = c; break }
      }

      colorMap[v] = color
      const regStr = color >= 0 ? regNames[color] || `R${color + 1}` : "SPILL"

      if (color >= 0) {
        // Optimistic success — even a "candidate" got colored
        steps.push({
          type: "ASSIGN",
          variable: v,
          registerId: color,
          message: `ASSIGN ${v} → ${regStr}${optimisticCandidates.has(v) ? " (optimistically colored!)" : ""}`,
          currentAssignment: { ...colorMap },
          currentStack: [],
          spillCandidates: [...spillCandidates],
          remainingGraph: {},
        })
      } else {
        spillCandidates.push(v)
        steps.push({
          type: "SPILL_DETECTED",
          variable: v,
          registerId: -1,
          message: `SPILL ${v} — even optimistic coloring failed`,
          currentAssignment: { ...colorMap },
          currentStack: [],
          spillCandidates: [...spillCandidates],
          remainingGraph: {},
        })
      }

      assignment[v] = {
        register: color,
        name: regStr,
      }
    }

    const usedRegs = new Set(Object.values(colorMap).filter(c => c >= 0))

    return {
      mode: "OPTIMISTIC",
      label: "Optimistic Coloring",
      spillCount: spillCandidates.length,
      stepCount: steps.length,
      registerUtilization: usedRegs.size / K,
      assignment,
      spillCandidates,
      steps,
    }
  }
}
